include/mockturtle/algorithms/node_resynthesis/xmg_npn.hpp

```diff
@@ -0,0 +1,293 @@
+    
+    auto func_str = "0x" + kitty::to_hex( std::get<0>( config ) ); 
+    const auto it = class2signal.find( func_str );
+    assert( it != class2signal.end() );
+
+    //const auto it = class2signal.find( static_cast<uint16_t>( std::get<0>( config )._bits[0] ) );
+
+    std::vector<xmg_network::signal> pis( 4, xmg.get_constant( false ) );
+    std::copy( begin, end, pis.begin() );
+
+    std::vector<xmg_network::signal> pis_perm( 4 );
+    auto perm = std::get<2>( config );
+    for ( auto i = 0; i < 4; ++i )
+    {
+      pis_perm[i] = pis[perm[i]];
+    }
+
+    const auto& phase = std::get<1>( config );
+    for ( auto i = 0; i < 4; ++i )
+    {
+      if ( ( phase >> perm[i] ) & 1 )
+      {
+        pis_perm[i] = !pis_perm[i];
+      }
+    }
+
+    for ( auto const& po : it->second )
+    {
+      topo_view topo{db, po};
+      auto f = cleanup_dangling( topo, xmg, pis_perm.begin(), pis_perm.end() ).front();
+
+      if ( !fn( ( ( phase >> 4 ) & 1 ) ? !f : f ) )
+      {
+        return; /* quit */
+      }
+    }
+  }
+
+private:
+  std::unordered_map<std::string, std::string> opt_xmgs;
+
+  inline std::vector<std::string> split( const std::string& str, const std::string& sep )
+  {
+    std::vector<std::string> result;
+
+    size_t last = 0;
+    size_t next = 0;
+    while ( ( next = str.find( sep, last ) ) != std::string::npos )
+    {
+      result.push_back( str.substr( last, next - last ) );
+      last = next + 1;
+    }
+    result.push_back( str.substr( last ) );
+
+    return result;
+  }
+
+  void load_optimal_xmgs( const unsigned& strategy )
+  {
+    std::vector<std::string> result;
+
+    switch( strategy )
+    {
+      case 1:
+        result = split( npn4_s, "\n" );
+        break;
+
+      case 2:
+        result = split( npn4_sd, "\n" );
+        break;
+      
+      case 3:
+        result = split( npn4_ds, "\n" );
+        break;
+
+      default: break;
+    }
+    
+    for( auto record : result )
+    {
+      auto p = split( record, " " );
+      assert( p.size() == 2u );
+      opt_xmgs.insert( std::make_pair( p[0], p[1] ) );
+    }
+
+  }
+
+  std::vector<xmg_network::signal> create_xmg_from_str( const std::string& str, const std::vector<xmg_network::signal>& signals )
+  {
+    auto sig = signals;
+    std::vector<xmg_network::signal> result;
+
+    std::stack<int> polar;
+    std::stack<xmg_network::signal> inputs;
+
+    for( auto i = 0ul; i < str.size(); i++ )
+    {
+      // operators polarity
+      if( str[i] == '[' || str[i] == '<' )
+      {
+        if( i == 0 )
+        {
+          polar.push( 0 );
+        }
+        else if( str[i - 1] == '!' )
+        {
+          polar.push( 1 );
+        }
+        else
+        {
+          polar.push( 0 );
+        }
+      }
+
+      //input signals
+      if( str[i] >= 'a' && str[i] <= 'd' ) 
+      {
+        inputs.push( sig[ str[i] - 'a' + 1 ] );
+
+        polar.push( str[i - 1] == '!' ? 1 : 0 );
```

Comment:
This causes a problem if `i == 0`, which is the case if `str == "a"`, for example. This buffer underflow is strangely only detected in the Win builds. I will merge and then fix this in `master`.